Enum Refactor: DataZoneEquipment

dareumnam opened the pull request:
Pull request overview
---------------------
Remove `constexpr int`s in DataZoneEquipment object

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

dareumnam committed:
DataZoneEquipment::ZoneEquip

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into enumDataZoneEquipment

dareumnam committed:
Change Array1D to std::array

dareumnam committed:
DataZoneEquipment::PerPersonVentRateMode

dareumnam committed:
enums start from -1

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into enumDataZoneEquipment

dareumnam committed:
fix unit tests

dareumnam committed:
Revert "fix unit tests"

This reverts commit 814a3b859b2f0fd058b6f3d21a52c8e951bfad18.

dareumnam committed:
Revert "enums start from -1"

This reverts commit 6cd174650929a1869cacd693dac3c2764b5ea5c6.

dareumnam committed:
Revert "DataZoneEquipment::PerPersonVentRateMode"

This reverts commit 895fae9c7bce606058ea587a9304851d02a2e958.

dareumnam committed:
Revert "Change Array1D to std::array"

This reverts commit feaf474e6eabdd1875a2782d13752f9047bba376.

dareumnam committed:
DataZoneEquipment::PerPersonVentRateMode

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into enumDataZoneEquipTest

amirroth commented on src/EnergyPlus/BaseboardElectric.cc:
```diff
@@ -338,7 +337,8 @@ namespace BaseboardElectric {
                 for (CtrlZone = 1; CtrlZone <= state.dataGlobal->NumOfZones; ++CtrlZone) {
                     for (ZoneEquipTypeNum = 1; ZoneEquipTypeNum <= state.dataZoneEquip->ZoneEquipList(CtrlZone).NumOfEquipTypes; ++ZoneEquipTypeNum) {
-                        if (state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipType_Num(ZoneEquipTypeNum) == BBElectricConvective_Num &&
+                        if (state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipType_Num(ZoneEquipTypeNum) ==
```
I know this is pedantic, but we should call things `Type` rather than `TypeNum`. Not so critical here, more important for member names.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
-    constexpr int ZoneMixer_Type(3);
-    constexpr int ZoneReturnPlenum_Type(4);
+    enum class CompType
+    {
```
Not sure `CompType` is a good name for this `enum`. Also, need to add `Num` to the ends of these.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
+    enum ZoneEquip
+    {
+        Unassigned = -1,
+        FanCoil4Pipe = 1,
```
Why is this not 0? Is it because there are `Array1D` that use these as indices? That should be a sign to switch those to `std::array`. Can do this in a subsequent PR I guess.

Myoldmopar replied:
@dareumnam any comment on this?

rraustad replied:
![image](https://user-images.githubusercontent.com/5778156/142266425-f67650a3-705b-4c28-88ab-8a0fcc861641.png)

amirroth replied:
Per my comment above, arrays indexed by `enum` should be turned into `std::array`. Per same comment, no need to hold up this PR for that single change.

Myoldmopar replied:
:+1:

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
+        RefrigerationAirChillerSet,
+        UserDefinedZoneHVACForcedAir,
+        CoolingPanel,
+        ZoneUnitarySys
```
Add `Num`.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
+        Unassgined = -1,
+        DCVByCurrentLevel,
+        ByDesignLevel
+    };
```
Add `Num`.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -358,7 +374,7 @@ namespace DataZoneEquipment {
         int NumAvailCoolEquip;                      // Number of pieces of equipment available for cooling
         Array1D_string EquipType;
-        Array1D_int EquipType_Num;
+        Array1D<DataZoneEquipment::ZoneEquip> EquipType_Num;
```
No `_Num`. I understand that there is already an array called `EquipType` which is an array of strings, but that array should not be necessary. You should convert things from string to `enum` as we parse the IDF and only use `enum` internally. If you want, you can put a pin in these comments and do them in a separate PR.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -407,7 +423,7 @@ namespace DataZoneEquipment {
         int InletNodeNum;
         Array1D_string ComponentType;
-        Array1D_int ComponentType_Num;
+        Array1D<DataZoneEquipment::CompType> ComponentType_Num;
```
Same comment as above.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -431,7 +447,7 @@ namespace DataZoneEquipment {
         int OutletNodeNum;
         Array1D_string ComponentType;
-        Array1D_int ComponentType_Num;
+        Array1D<DataZoneEquipment::CompType> ComponentType_Num;
```
Same comment.

amirroth commented on src/EnergyPlus/FanCoilUnits.cc:
```diff
@@ -712,7 +712,7 @@ namespace FanCoilUnits {
                        FanCoil(FanCoilNum).AirOutNode);
             FanCoil(FanCoilNum).ControlZoneNum =
-                DataZoneEquipment::GetZoneEquipControlledZoneNum(state, DataZoneEquipment::FanCoil4Pipe_Num, FanCoil(FanCoilNum).Name);
+                DataZoneEquipment::GetZoneEquipControlledZoneNum(state, DataZoneEquipment::ZoneEquip::FanCoil4Pipe, FanCoil(FanCoilNum).Name);
```
Probably not something you want to handle in this PR, but ... why are we doing this lookup by FanCoil name if we already know FanCoil index?

rraustad replied:
ZoneEquipConfig stores the component type and name so to look up the correct component needed the type and name (back then). Now that code is progressing a new method might be used.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -516,107 +516,107 @@ void GetZoneEquipmentData(EnergyPlusData &state)
                     auto const SELECT_CASE_var(UtilityRoutines::MakeUPPERCase(thisZoneEquipList.EquipType(ZoneEquipTypeNum)));
 
                     if (SELECT_CASE_var == "ZONEHVAC:AIRDISTRIBUTIONUNIT") {
```
This code should be converted to a `getEnumerationValue`.

amirroth commented on src/EnergyPlus/ReturnAirPathManager.cc:
```diff
@@ -191,9 +188,9 @@ namespace ReturnAirPathManager {
                         }
                         if (UtilityRoutines::SameString(state.dataIPShortCut->cAlphaArgs(Counter), "AirLoopHVAC:ZoneMixer"))
-                            state.dataZoneEquip->ReturnAirPath(PathNum).ComponentType_Num(CompNum) = ZoneMixer_Type;
+                            state.dataZoneEquip->ReturnAirPath(PathNum).ComponentType_Num(CompNum) = DataZoneEquipment::CompType::ZoneMixer;
```
`getEnumerationValue`.

amirroth commented on src/EnergyPlus/SystemReports.cc:
```diff
@@ -4539,7 +4539,7 @@ void ReportMaxVentilationLoads(EnergyPlusData &state)
                 // case statement to cover all possible zone forced air units that could have outside air
 
-                if (SELECT_CASE_var == WindowAC_Num) { // Window Air Conditioner
+                if (SELECT_CASE_var == DataZoneEquipment::ZoneEquip::WindowAC) { // Window Air Conditioner
```
This should be a `switch`/`case`.

amirroth commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -3065,7 +3065,7 @@ void SimZoneEquipment(EnergyPlusData &state, bool const FirstHVACIteration, bool
                 auto const SELECT_CASE_var(state.dataZoneEquip->SupplyAirPath(SupplyAirPathNum).ComponentType_Num(CompNum));
 
-                if (SELECT_CASE_var == ZoneSplitter_Type) { // 'AirLoopHVAC:ZoneSplitter'
+                if (SELECT_CASE_var == DataZoneEquipment::CompType::ZoneSplitter) { // 'AirLoopHVAC:ZoneSplitter'
```
`switch`/`case`.

dareumnam committed:
clang format

dareumnam committed:
address comments

dareumnam committed:
address comments

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix errors caused by renaming

dareumnam committed:
fix errors caused by renaming 2

dareumnam committed:
fix errors caused by renaming 3

dareumnam committed:
fix errors caused by renaming 4

dareumnam committed:
fix errors caused by renaming 5

Myoldmopar commented:
Looks like comments have been addressed here, but speak up if there is anything still waiting.  I pulled develop into this locally and re-ran regressions.  Everything passes beautifully.  If no one has an issue, I'll merge this this afternoon.

Myoldmopar commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -12325,9 +12325,6 @@ namespace AirflowNetworkBalanceManager {
         // This subroutine validate zone exhaust fan and associated surface
 
-        // Using/Aliasing
-        using DataZoneEquipment::ZoneExhaustFan_Num;
```
A simple but nice change.

Myoldmopar commented on src/EnergyPlus/ConvectionCoefficients.cc:
```diff
@@ -5823,9 +5823,9 @@ void DynamicIntConvSurfaceClassification(EnergyPlusData &state, int const SurfNu
-                case PurchasedAir_Num:
+                switch (state.dataZoneEquip->ZoneEquipList(state.dataZoneEquip->ZoneEquipConfig(ZoneNum).EquipListIndex).EquipTypeEnum(EquipNum)) {
+                case DataZoneEquipment::ZoneEquip::AirDistUnit:
+                case DataZoneEquipment::ZoneEquip::PurchasedAir:
```
This reads so much nicer.

Myoldmopar commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -76,63 +76,87 @@ namespace DataZoneEquipment {
+        Intermediate,
+        Outlet,
+        NUM
+    };
```
Awesome.

Myoldmopar reviewed (approved):
Everything looks OK to me.  There appears to be one unanswered question which I noted in my comments, but even that is something for later.  I think this is good to go.

dareumnam commented:
@Myoldmopar Thanks for your review. I keep addressing comments on my local branch, but it is okay if this gets merged in first. I can open a separate PR.

Myoldmopar commented:
Alright, pulled develop in just for fun after the version change and it's still passing fine.  @dareumnam just make sure to address the last little pieces here on your next PR.  Thanks!  Merging!
